vidlistener/listener.go

```diff
@@ -17,38 +22,90 @@ type VidListener struct {
+//HandleRTMPPublish immediately turns the RTMP stream into segmented HLS, and writes it into a stream.
+//It exposes getStreamID so the user can name the stream, and getStream so the user can keep track of all the streams.
+func (self *VidListener) HandleRTMPPublish(
 	getStreamID func(reqPath string) (string, error),
-	getStream func(reqPath string) (stream.Stream, error),
+	getStream func(reqPath string) (stream.Stream, stream.Stream, error),
 	endStream func(reqPath string)) error {
 
-	s.RtmpServer.HandlePublish = func(conn *joy4rtmp.Conn) {
+	self.RtmpServer.HandlePublish = func(conn *joy4rtmp.Conn) {
 		glog.Infof("RTMP server got upstream")
 
-		streamID, err := getStreamID(conn.URL.Path)
+		_, err := getStreamID(conn.URL.Path)
 		if err != nil {
 			glog.Errorf("RTMP Stream Publish Error: %v", err)
 			return
 		}
 
-		stream, err := getStream(conn.URL.Path)
+		rs, hs, err := getStream(conn.URL.Path)
 		if err != nil {
 			glog.Errorf("RTMP Publish couldn't get a destination stream for %v", conn.URL.Path)
 			return
 		}
+		// hlsS := stream.NewVideoStream(rs.GetStreamID() + ".m3u8")
+
+		glog.Infof("Got RTMP Stream: %v", rs.GetStreamID())
+		cew := make(chan error, 0)
+		cs := make(chan error, 0)
+
+		glog.Infof("Writing RTMP to stream")
+		go func() { cew <- rs.WriteRTMPToStream(context.Background(), conn) }()
+		go func() { cs <- self.segmentStream(rs, hs) }()
 
-		glog.Infof("Got RTMP Stream: %v", streamID)
-		c := make(chan error, 0)
-		go func() { c <- stream.WriteRTMPToStream(context.Background(), conn) }()
 		select {
-		case err := <-c:
+		case err := <-cew:
 			endStream(conn.URL.Path)
-			glog.Infof("Final stream length: %v", stream.Len())
+			glog.Infof("Final stream length: %v", rs.Len())
 			glog.Error("Got error writing RTMP: ", err)
+		case err := <-cs:
+			glog.Errorf("Error segmenting, %v", err)
 		}
 
 	}
 	return nil
 }
+
+func (self *VidListener) segmentStream(rs stream.Stream, hs stream.Stream) error {
+	// //Invoke Segmenter
+	workDir, _ := os.Getwd()
+	workDir = workDir + "/tmp"
+	localRtmpUrl := "rtmp://localhost" + self.RtmpServer.Addr + "/stream/" + rs.GetStreamID()
+	ctx := context.Background()
+	s := segmenter.NewFFMpegVideoSegmenter(workDir, rs.GetStreamID(), localRtmpUrl, segOptions.SegLength)
+	c := make(chan error, 1)
+	go func() { c <- s.RTMPToHLS(ctx, segOptions) }()
+
+	go func() {
+		c <- func() error {
+			for {
+				pl, err := s.PollPlaylist(ctx)
+				if err != nil {
+					glog.Errorf("Got error polling playlist: %v", err)
+					return err
+				}
+				// glog.Infof("Writing pl: %v", pl)
+				hs.WriteHLSPlaylistToStream(*pl.Data)
+			}
+		}()
+	}()
+
+	go func() {
+		c <- func() error {
+			for {
+				seg, err := s.PollSegment(ctx)
+				if err != nil {
+					return err
+				}
+				ss := stream.HLSSegment{Data: seg.Data, Name: seg.Name}
+				glog.Infof("Writing stream: %v, len:%v", ss.Name, len(seg.Data))
+				hs.WriteHLSSegmentToStream(ss)
+			}
+		}()
+	}()
+	select {
+	case err := <-c:
```

Comment:
Can this listener be killed in any way, or does it just block forever? Or does PollSegment always eventually return an error?